### PR TITLE
#15498: changed read and compute kernels for height reduction

### PR DESCRIPTION
**Pull request by ipotkonjak-tt**
### Ticket
Issue https://github.com/tenstorrent/tt-metal/issues/15498

### Problem description
Problem appears in height reduction when input is DRAM interleaved. In this case tensor width directly impacts which DRAM banks will cores visit when reading column tiles. In case of WH the number of banks can range from 1 to 12, where the optimal case is for all 12 to be utilised, while read requests are distributed evenly across them. As tensor shape is an input parameter height reduction kernels need to be modified to support the presence of different reading patterns.

### What's changed
Read and compute kernels for height interleaved reduction were adapted for column reading. Instead of processing each column separately, columns that are assigned to one core are split into batches and processed row by row. This is done to mitigate the variability introduced by different DRAM reading patterns, by increasing the number of utilised banks per one processing cycle.

### Checklist
- [x] Post commit CI passes
- [ ] Blackhole Post commit
- [x] Model regression CI testing passes* 
- [x] Device performance regression CI testing passes
- [x] New/Existing tests provide coverage for changes

*except for the tests failing on main

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/reduction/generic/device/multi_core_h/reduce_op_multi_core_h.cpp [hunk 1]
```diff
@@ -119,6 +119,10 @@ operation::ProgramWithCallbacks reduce_multi_core_h(
     tt_metal::KernelHandle reader_kernel_id;
     bfloat16 bfloat_scaler_value = bfloat16(scaler);
     uint32_t packed_scaler_value = pack_two_bfloat16_into_uint32({bfloat_scaler_value, bfloat_scaler_value});
+
+    uint32_t chunk_size =
```
[the remark this answers is not in the document]
Comment (ipotkonjak-tt) [reply ttnn/cpp/ttnn/operations/reduction/generic/device/multi_core_h/reduce_op_multi_core_h.cpp:123]: Made the change, `ttnn::get_dest_reg_count()` is used.

File: ttnn/cpp/ttnn/operations/reduction/generic/device/kernels/compute/reduce_h_interleaved.cpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    cb_wait_front(tt::CB::c_in2, 1);  // scaler tile from the reader
+
+    constexpr int onetile = 1;
+    for (uint32_t nc = 0; nc < NC; ++nc) {
+        for (uint32_t wt = 0; wt < Wt; wt += row_chunk) {
+            uint32_t chunk_end = std::min(wt + row_chunk, Wt);
+            uint32_t tile_num = std::min(row_chunk, Wt - wt);
+            int reduce_dst_idx = 0;
+
+            // reduction for a chunk of columns
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/reduction/generic/device/kernels/compute/reduce_h_interleaved.cpp:27]: Please add a long comment about the order the values are coming in, similar to the old kernel, except with more detail.

Need to include details about
- N C W_skip H W_chunk order
- say what the difference is from NCWH order 
- reword original last comment line to something like "in this case we just sequentially add to accumulator all the H-tiles in a column, with each entry of a row chunk uses a different dest index"
With an example (Ht=3, Wt=4, row_chunk=2 is the simplest I could think of), showing wt,ht indices. I'm not sure if this should be triplets (with reduce_dst_index being the third entry, although that seems like too much information).
```
      Wt
Ht 0,0 0,0 1,0 1,0
     0,1. 0,1 1,1 1,1
     0,2. 0,2 1,1 1,1
```

File: ttnn/cpp/ttnn/operations/reduction/generic/device/kernels/dataflow/reader_unary_transpose_wh_interleaved_input_cols_partitioned.cpp [hunk 3]
```diff
@@ -36,23 +37,38 @@ void kernel_main() {
 
     uint32_t w = curr_col_in_batch;
 
-    // this reader will read a NHW tensor in NWH order
-    for (uint32_t i = 0; i < num_cols; i++) {
+    for (uint32_t i = 0; i < num_cols; i += row_chunk) {
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/reduction/generic/device/kernels/dataflow/reader_unary_transpose_wh_interleaved_input_cols_partitioned.cpp:40]: Please add a long comment about how this reads in data.
Need to include details about
- read a NHW tensor in N W_skip H W_chunk order
say what the difference is from NWH order
With an example (Ht=3, Wt=4, row_chunk=2 is the simplest I could think of), showing order of reading:
```
      Wt
Ht 0 1 6 7
     2 3 8 9
     4 5 10 11
```

File: ttnn/cpp/ttnn/operations/reduction/generic/device/kernels/dataflow/reader_unary_transpose_wh_interleaved_input_cols_partitioned.cpp [hunk 4]
```diff
@@ -36,23 +37,38 @@ void kernel_main() {
...
-        if (w == Wt) {
-            col_start_tile_id = curr_id - Wt + 1;
-            w = 0;
-        } else {
-            col_start_tile_id++;
+        uint32_t reset_curr_id = curr_id;
+        uint32_t reset_w = w;
+        uint32_t reset_col_start = col_start_tile_id;
+
+        // row wise read for a chunk of columns
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/reduction/generic/device/kernels/dataflow/reader_unary_transpose_wh_interleaved_input_cols_partitioned.cpp:47]: Please add comments about how the reset variables work. This logic is complicated.

File: ttnn/cpp/ttnn/operations/reduction/generic/device/kernels/compute/reduce_h_interleaved.cpp [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+            for (uint32_t ht = 0; ht < Ht; ++ht) {
+                reduce_dst_idx = 0;
+                for (uint32_t i = wt; i < chunk_end; ++i) {
+                    cb_wait_front(tt::CB::c_in0, onetile);
+                    reduce_tile(tt::CB::c_in0, tt::CB::c_in2, 0, 0, reduce_dst_idx);
+                    cb_pop_front(tt::CB::c_in0, onetile);
+                    ++reduce_dst_idx;
+                }
+            }
+            for (uint32_t i = 0; i < tile_num; i++) {
```
[the remark this answers is not in the document]
Comment (ipotkonjak-tt) [reply ttnn/cpp/ttnn/operations/reduction/generic/device/kernels/compute/reduce_h_interleaved.cpp]: tile_num removed, loop is now as reduce_tile one.

File: ttnn/cpp/ttnn/operations/reduction/generic/device/kernels/compute/reduce_h_interleaved.cpp (hunk 2 above)
Comment (ipotkonjak-tt) [reply ttnn/cpp/ttnn/operations/reduction/generic/device/kernels/compute/reduce_h_interleaved.cpp]: Added a detailed comment with the example, hopefully didn't go overboard.

File: ttnn/cpp/ttnn/operations/reduction/generic/device/kernels/dataflow/reader_unary_transpose_wh_interleaved_input_cols_partitioned.cpp (hunk 4 above)
Comment (ipotkonjak-tt) [reply ttnn/cpp/ttnn/operations/reduction/generic/device/kernels/dataflow/reader_unary_transpose_wh_interleaved_input_cols_partitioned.cpp]: Explained the reasoning behind reset variables, with comments on each one individually.

File: ttnn/cpp/ttnn/operations/reduction/generic/device/kernels/compute/reduce_h.cpp [hunk 6]
```diff
@@ -11,28 +11,44 @@ void MAIN {
     uint32_t Ht = get_compile_time_arg_val(0);
     uint32_t Wt = get_compile_time_arg_val(1);
     uint32_t NC = get_compile_time_arg_val(2);
+    uint32_t row_chunk = get_compile_time_arg_val(3);
 
-    reduce_init<true>(tt::CBIndex::c_0, tt::CBIndex::c_2);
-    cb_wait_front(tt::CBIndex::c_2, 1);  // scaler tile from the reader
+    reduce_init<true>(tt::CB::c_in0, tt::CB::c_in2);
+    cb_wait_front(tt::CB::c_in2, 1);  // scaler tile from the reader
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/reduction/generic/device/kernels/compute/reduce_h.cpp:17]: With https://github.com/tenstorrent/tt-metal/pull/14971 we no longer have restrictions on CBs. Therefore we should move away from c_in0 etc. and use c_0 etc.

File: ttnn/cpp/ttnn/operations/reduction/generic/device/kernels/dataflow/reader_unary_transpose_wh_interleaved_input_cols_partitioned.cpp [hunk 7]
```diff
@@ -17,6 +17,7 @@ void kernel_main() {
     constexpr uint32_t Ht = get_compile_time_arg_val(1);
     constexpr uint32_t Wt = get_compile_time_arg_val(2);
     constexpr uint32_t HtWt = get_compile_time_arg_val(3);
+    constexpr uint32_t row_chunk = get_compile_time_arg_val(4);
 
     constexpr uint32_t cb_id_in0 = 0;
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/reduction/generic/device/kernels/dataflow/reader_unary_transpose_wh_interleaved_input_cols_partitioned.cpp:22]: It'd be good to change 0 to tt::CBIndex::c_0 and the 2 on line 29 below to tt::CBIndex::c_2

**Review by bbradelTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/reduction/generic/device/kernels/compute/reduce_h.cpp (hunk 6 above)
Comment (ipotkonjak-tt) [reply ttnn/cpp/ttnn/operations/reduction/generic/device/kernels/compute/reduce_h.cpp]: Replaced with `tt::CBIndex::c_0` and `tt::CBIndex::c_2`

**Review by bbradelTT — APPROVED** [review]

**Pull request merged**